File: web/src/data/disputes.ts
/** Disputes, from either side of one and from the moderator's chair. */

import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'

import { api } from '@/data/client'
import { JOBS_KEY } from '@/data/jobs'
import type { Page } from '@/data/types'

export type DisputeReason =
  | 'no_show'
  | 'work_not_done'
  | 'damage'
  | 'price_disagreement'
  | 'behaviour'
  | 'other'

export type DisputeStatus = 'open' | 'claimed' | 'resolved'
export type DisputeVerdict = 'client_at_fault' | 'provider_at_fault' | 'no_fault'

export const DISPUTE_REASONS: DisputeReason[] = [
  'no_show',
  'work_not_done',
  'damage',
  'price_disagreement',
  'behaviour',
  'other',
]

export interface DisputeParty {
  id: number
  full_name: string
  role: string
  provider_id: number | null
  rating_avg: number | null
  rating_count: number | null
  jobs_done: number | null
  disputes_lost: number
}

export interface DisputeMessage {
  id: number
  author_id: number
  author_name: string
  body: string
  attachment_url: string | null
  is_internal: boolean
  created_at: string
}

export interface DisputeJob {
  id: number
  title: string
  status: string
  agreed_price_centimes: number
  finished_at: string | null
  /** The only money figure a moderator is given, because he can refund it. */
  lead_fee_centimes: number | null
}

export interface Dispute {
  id: number
  reason: DisputeReason
  description: string
  status: DisputeStatus
  created_at: string
  job: DisputeJob
  opened_by: DisputeParty
  against: DisputeParty
  claimed_by_id: number | null
  claimed_by_name: string | null
  claimed_at: string | null
  verdict: DisputeVerdict | null
  resolution_note: string | null
  lead_fee_refunded: boolean
  resolved_at: string | null
  evidence: string[]
  messages: DisputeMessage[]
  is_stale: boolean
}

export interface DisputeRow {
  id: number
  reason: DisputeReason
  status: DisputeStatus
  created_at: string
  job_title: string
  opened_by_name: string
  against_name: string
  claimed_by_id: number | null
  is_stale: boolean
}

export type QueueTab = 'open' | 'mine' | 'resolved'

export const DISPUTES_KEY = ['disputes'] as const
export const QUEUE_KEY = ['mod', 'disputes'] as const

export function useMyDisputes() {
  return useQuery({
    queryKey: DISPUTES_KEY,
    queryFn: () => api<Page<DisputeRow>>('/disputes?per_page=50'),
    staleTime: 15_000,
  })
}

export function useDispute(disputeId: number | null) {
  return useQuery({
    queryKey: [...DISPUTES_KEY, disputeId],
    queryFn: () => api<Dispute>(`/disputes/${disputeId}`),
    enabled: disputeId !== null,
    staleTime: 10_000,
  })
}

export function useDisputeQueue(tab: QueueTab) {
  return useQuery({
    queryKey: [...QUEUE_KEY, tab],
    queryFn: () => api<Page<DisputeRow>>(`/mod/disputes?tab=${tab}&per_page=50`),
    staleTime: 10_000,
  })
}

/** Everything a dispute action touches: the case, both queues, and the job. */
function useDisputeMutation<TVariables>(request: (variables: TVariables) => Promise<Dispute>) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: request,
    onSuccess: (dispute) => {
      queryClient.setQueryData([...DISPUTES_KEY, dispute.id], dispute)
      void queryClient.invalidateQueries({ queryKey: DISPUTES_KEY })
      void queryClient.invalidateQueries({ queryKey: QUEUE_KEY })
      void queryClient.invalidateQueries({ queryKey: JOBS_KEY })
    },
  })
}

export function useOpenDispute() {
  return useDisputeMutation(
    ({
      jobId,
      ...body
    }: {
      jobId: number
      reason: DisputeReason
      description: string
      evidence_paths: string[]
    }) => api<Dispute>(`/jobs/${jobId}/dispute`, { method: 'POST', body }),
  )
}

export function useClaimDispute() {
  return useDisputeMutation((disputeId: number) =>
    api<Dispute>(`/mod/disputes/${disputeId}/claim`, { method: 'POST' }),
  )
}

export function useDisputeMessage() {
  return useDisputeMutation(
    ({ disputeId, body, isInternal }: { disputeId: number; body: string; isInternal?: boolean }) =>
      api<Dispute>(`/disputes/${disputeId}/messages`, {
        method: 'POST',
        body: { body, is_internal: isInternal ?? false },
      }),
  )
}

export function useResolveDispute() {
  return useDisputeMutation(
    ({
      disputeId,
      ...body
    }: {
      disputeId: number
      verdict: DisputeVerdict
      note: string
      refund_lead_fee: boolean
      suspend_at_fault: boolean
    }) => api<Dispute>(`/mod/disputes/${disputeId}/resolve`, { method: 'POST', body }),
  )
}

File: api/app/api/routes/disputes.py
"""C8, D1 and D2 — opening a dispute, the queue, and settling one."""

from __future__ import annotations

from typing import Annotated, Literal

from fastapi import APIRouter, Depends, Query, Request, status

from app.core.dispute import is_stale
from app.core.enums import Role
from app.deps import CurrentUser, DbSession, require_roles
from app.models.base import utcnow
from app.models.dispute import Dispute
from app.models.offer import Offer
from app.models.user import User
from app.schemas.common import Page
from app.schemas.dispute import (
    DisputeJobOut,
    DisputeMessageIn,
    DisputeMessageOut,
    DisputeOut,
    DisputePartyOut,
    DisputeRowOut,
    NewDisputeIn,
    ResolveDisputeIn,
)
from app.services.disputes import DisputeService

router = APIRouter(tags=["disputes"])

Staff = Depends(require_roles(Role.MODERATOR, Role.ADMIN))
Party = Depends(require_roles(Role.CLIENT, Role.PROVIDER))


@router.post(
    "/jobs/{job_id}/dispute",
    response_model=DisputeOut,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Party],
)
def open_dispute(
    job_id: int, payload: NewDisputeIn, user: CurrentUser, db: DbSession
) -> DisputeOut:
    """C8."""
    service = DisputeService(db)
    return _out(service, service.open(user, job_id, payload), user)


@router.get("/disputes", response_model=Page[DisputeRowOut], dependencies=[Party])
def my_disputes(
    user: CurrentUser,
    db: DbSession,
    page: Annotated[int, Query(ge=1)] = 1,
    per_page: Annotated[int, Query(ge=1, le=50)] = 20,
) -> Page[DisputeRowOut]:
    service = DisputeService(db)
    rows, total = service.mine(user, page=page, per_page=per_page)
    return Page[DisputeRowOut](
        items=[_row(service, dispute) for dispute in rows],
        total=total,
        page=page,
        per_page=per_page,
    )


@router.get("/mod/disputes", response_model=Page[DisputeRowOut], dependencies=[Staff])
def dispute_queue(
    user: CurrentUser,
    db: DbSession,
    tab: Literal["open", "mine", "resolved"] = "open",
    page: Annotated[int, Query(ge=1)] = 1,
    per_page: Annotated[int, Query(ge=1, le=50)] = 20,
) -> Page[DisputeRowOut]:
    """D1. Oldest first — two people are waiting on each row."""
    service = DisputeService(db)
    rows, total = service.queue(user, tab=tab, page=page, per_page=per_page)
    return Page[DisputeRowOut](
        items=[_row(service, dispute) for dispute in rows],
        total=total,
        page=page,
        per_page=per_page,
    )


@router.get("/disputes/{dispute_id}", response_model=DisputeOut)
def get_dispute(dispute_id: int, user: CurrentUser, db: DbSession) -> DisputeOut:
    """D2 for staff; the same case, minus the internal notes, for a party."""
    service = DisputeService(db)
    return _out(service, service.get(user, dispute_id), user)


@router.post("/mod/disputes/{dispute_id}/claim", response_model=DisputeOut, dependencies=[Staff])
def claim_dispute(dispute_id: int, user: CurrentUser, db: DbSession) -> DisputeOut:
    service = DisputeService(db)
    return _out(service, service.claim(user, dispute_id), user)


@router.post(
    "/disputes/{dispute_id}/messages",
    response_model=DisputeOut,
    status_code=status.HTTP_201_CREATED,
)
def add_message(
    dispute_id: int, payload: DisputeMessageIn, user: CurrentUser, db: DbSession
) -> DisputeOut:
    service = DisputeService(db)
    service.add_message(user, dispute_id, payload)
    return _out(service, service.get(user, dispute_id), user)


@router.post(
    "/mod/disputes/{dispute_id}/resolve", response_model=DisputeOut, dependencies=[Staff]
)
def resolve_dispute(
    dispute_id: int,
    payload: ResolveDisputeIn,
    user: CurrentUser,
    db: DbSession,
    request: Request,
) -> DisputeOut:
    service = DisputeService(db)
    resolved = service.resolve(user, dispute_id, payload, ip=_ip(request))
    return _out(service, resolved, user)


def _row(service: DisputeService, dispute: Dispute) -> DisputeRowOut:
    job = service.repo.job_for(dispute)
    request = service.jobs.request_for(job)
    opener = service.repo.user(dispute.opened_by_id)
    against = service.repo.user(dispute.against_id)

    return DisputeRowOut(
        id=dispute.id,
        reason=dispute.reason,
        status=dispute.status,
        created_at=dispute.created_at,
        job_title=request.title,
        opened_by_name=opener.full_name if opener else "—",
        against_name=against.full_name if against else "—",
        claimed_by_id=dispute.claimed_by_id,
        is_stale=dispute.claimed_by_id is None and is_stale(dispute.created_at, utcnow()),
    )


def _out(service: DisputeService, dispute: Dispute, reader: User) -> DisputeOut:
    job = service.repo.job_for(dispute)
    request = service.jobs.request_for(job)
    offer = service.db.get(Offer, job.offer_id)

    opener = service.repo.user(dispute.opened_by_id)
    against = service.repo.user(dispute.against_id)
    claimed_by = (
        service.repo.user(dispute.claimed_by_id) if dispute.claimed_by_id else None
    )

    messages = service.visible_messages(reader, dispute)
    names = {
        user.id: user.full_name
        for user in (opener, against, claimed_by)
        if user is not None
    }

    return DisputeOut(
        id=dispute.id,
        reason=dispute.reason,
        description=dispute.description,
        status=dispute.status,
        created_at=dispute.created_at,
        job=DisputeJobOut(
            id=job.id,
            title=request.title,
            status=job.status.value,
            agreed_price_centimes=job.agreed_price_centimes,
            finished_at=job.finished_at,
            # The one money figure a moderator sees, because he can refund it.
            lead_fee_centimes=offer.lead_fee_centimes if offer else None,
        ),
        opened_by=_party(service, opener),
        against=_party(service, against),
        claimed_by_id=dispute.claimed_by_id,
        claimed_by_name=claimed_by.full_name if claimed_by else None,
        claimed_at=dispute.claimed_at,
        verdict=dispute.verdict,
        resolution_note=dispute.resolution_note,
        lead_fee_refunded=dispute.lead_fee_refunded,
        resolved_at=dispute.resolved_at,
        evidence=[
            message.attachment_url
            for message in dispute.messages
            if message.attachment_url
        ],
        messages=[
            DisputeMessageOut(
                id=message.id,
                author_id=message.author_id,
                author_name=names.get(message.author_id, "—"),
                body=message.body,
                attachment_url=message.attachment_url,
                is_internal=message.is_internal,
                created_at=message.created_at,
            )
            for message in messages
        ],
        is_stale=dispute.claimed_by_id is None and is_stale(dispute.created_at, utcnow()),
    )


def _party(service: DisputeService, user: User | None) -> DisputePartyOut:
    if user is None:
        return DisputePartyOut(id=0, full_name="—", role="deleted")

    profile = service.repo.profile_for_user(user.id)
    return DisputePartyOut(
        id=user.id,
        full_name=user.full_name,
        role=user.role.value,
        provider_id=profile.id if profile else None,
        rating_avg=profile.rating_avg if profile else None,
        rating_count=profile.rating_count if profile else None,
        jobs_done=profile.jobs_done if profile else None,
        disputes_lost=service.repo.disputes_lost(user.id, user.role),
    )


def _ip(request: Request) -> str | None:
    return request.client.host if request.client else None
